### PR TITLE
[Task 6] As an user I can submit my reports on Slack

### PR DESCRIPTION
**Pull request by noor-ahmed**
## What happened
[Task-6](https://github.com/nimblehq/reportedly-api/projects/1#card-50159278)
 
## Insight
- [x] Set slack hook for posting report in slack
       - This will be under a thread
       - one message per day, reports will be inside the threads  

## Proof Of Work
 
Screenshot 2020-12-11 at 9 31 32 AM

### CODE REVIEW

File: app/models/report.rb [hunk 1]
```diff
@@ -1,7 +1,43 @@
...
   belongs_to :user
 
   validates :task_today, :obstacles_today, presence: true
+
+  def push_to_slack
+    thread_ts = SlackThread.find_by(created_on: Time.zone.today).thread_ts
+
+    uri = URI.parse('https://slack.com/api/chat.postMessage')
+
+    req_options = { use_ssl: uri.scheme == 'https' }
```
Comment (rafayetn) [inline app/models/report.rb:16]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/nimblehq/reportedly-api/blob/fb89d1c90ff855d229a9544a010555f929601d1c/app/models/report.rb#L15"></span>Refers to 'uri' more than self (maybe move it to another class?)</td>
    </tr>
  </tbody>
</table>

File: app/models/report.rb [hunk 2]
```diff
@@ -1,7 +1,43 @@
...
+
+    Net::HTTP.start(uri.hostname, uri.port, req_options) do |http|
```
Comment (rafayetn) [inline app/models/report.rb:18]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/nimblehq/reportedly-api/blob/fb89d1c90ff855d229a9544a010555f929601d1c/app/models/report.rb#L17"></span>Refers to 'uri' more than self (maybe move it to another class?)</td>
    </tr>
  </tbody>
</table>

File: app/models/report.rb [hunk 3]
```diff
@@ -1,7 +1,43 @@
...
+      http.request(set_request_body(uri, thread_ts))
+    end
+  end
+
+  private
+
+  def set_request_body(uri, thread_ts)
+    request = Net::HTTP::Post.new(uri)
+    request.content_type = 'application/json'
```
Comment (rafayetn) [inline app/models/report.rb:28]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/nimblehq/reportedly-api/blob/fb89d1c90ff855d229a9544a010555f929601d1c/app/models/report.rb#L26"></span>Refers to 'request' more than self (maybe move it to another class?)</td>
    </tr>
  </tbody>
</table>

File: app/models/report.rb [hunk 4]
```diff
@@ -1,7 +1,43 @@
...
+    request['Authorization'] = ENV['SLACK_REPLY_TOKEN']
```
Comment (rafayetn) [inline app/models/report.rb:29]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/nimblehq/reportedly-api/blob/fb89d1c90ff855d229a9544a010555f929601d1c/app/models/report.rb#L27"></span>Refers to 'request' more than self (maybe move it to another class?)</td>
    </tr>
  </tbody>
</table>

File: app/models/report.rb [hunk 5]
```diff
@@ -1,7 +1,43 @@
...
+    request.body = JSON.dump({
```
Comment (rafayetn) [inline app/models/report.rb:30]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/nimblehq/reportedly-api/blob/fb89d1c90ff855d229a9544a010555f929601d1c/app/models/report.rb#L28"></span>Refers to 'request' more than self (maybe move it to another class?)</td>
    </tr>
  </tbody>
</table>

File: app/services/slack_notification_service.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+require 'json'
+
+class SlackNotificationService
+  class << self
+    def init_reporting_thread
+      uri = URI.parse('https://slack.com/api/chat.postMessage')
+
+      req_options = { use_ssl: uri.scheme == 'https' }
+
+      return unless SlackThread.find_by(created_on: Time.zone.today).nil?
```
Comment (rafayetn) [inline app/services/slack_notification_service.rb:14]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/nimblehq/reportedly-api/blob/fb89d1c90ff855d229a9544a010555f929601d1c/app/services/slack_notification_service.rb#L14"></span>Calls 'time.zone' 2 times</td>
    </tr>
  </tbody>
</table>

Comment (rafayetn) [inline app/services/slack_notification_service.rb:14]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/nimblehq/reportedly-api/blob/fb89d1c90ff855d229a9544a010555f929601d1c/app/services/slack_notification_service.rb#L14"></span>Calls 'time.zone.today' 2 times</td>
    </tr>
  </tbody>
</table>

Comment (rafayetn) [inline app/services/slack_notification_service.rb:14]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/nimblehq/reportedly-api/blob/fb89d1c90ff855d229a9544a010555f929601d1c/app/services/slack_notification_service.rb#L14"></span>Performs a nil-check</td>
    </tr>
  </tbody>
</table>

File: app/services/slack_notification_service.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+      result = slack_result(uri, req_options)
+
+      SlackThread.create(thread_ts: result['ts'], created_on: Time.zone.today)
```
Comment (rafayetn) [inline app/services/slack_notification_service.rb:18]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/nimblehq/reportedly-api/blob/fb89d1c90ff855d229a9544a010555f929601d1c/app/services/slack_notification_service.rb#L18"></span>Calls 'time.zone' 2 times</td>
    </tr>
  </tbody>
</table>

Comment (rafayetn) [inline app/services/slack_notification_service.rb:18]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/nimblehq/reportedly-api/blob/fb89d1c90ff855d229a9544a010555f929601d1c/app/services/slack_notification_service.rb#L18"></span>Calls 'time.zone.today' 2 times</td>
    </tr>
  </tbody>
</table>

**Comment by rafayetn** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Code coverage is now at 0.00% (0/240 lines)</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by rafayet-monon — APPROVED** [review]
Initially approved!

**Pull request merged**